### PR TITLE
Improved Haskell language layer README.

### PR DESCRIPTION
**Pull request by Martinsos**
Improved Haskell README as discussed here https://github.com/syl20bnr/spacemacs/issues/13869 @smile13241324 .

There are a few things I was wondering about:
1. groovy layer README does not mention that user should enable auto-complete or syntax-checking layer. I wonder why is that so?
2. Syntax-checking is still a standalone section. I was considering if I can make it part of Configuration, but it really goes beyond just configuration, it explains how things work, so I decided to leave it as it is for now.
3. I wanted to make it very obvious what the minimal setup is. I am not sure if I managed to do that. Right now, I think new user needs to read Description, which is short, Install, which is straight forward, and then Configuration, and they are ready to go. Key bindings and especially Syntax checking and FAQ are kind of like bonuses and are not required to get started. But is that clear to the user? So I was thinking, maybe we could have a "QuickStart" section at the beginning that would summarize only the critical information from Description, Install and Configuration? But I gave up on that because it still felt like a lot of duplication. I might experiment more with this if I manage to get my brother to use Spacemacs :D, I can observe him then and see where he will get stuck. I am driving this whole point because I felt overwhelmed when I first started reading Haskell README, but now I know it very well so I can't remember exactly what I was confused about.

EDIT: Actually, I just added "Quick start" section which I think could solve the problem I described above under point number three -> it very quickly instructs user how to move around the README and what to look at first. Let me know what you think about it, we can remove it if it is too much. Btw in this Quick start section I tried using internal links to headers in Org, and they worked in emacs, but they don't work when README is read via Github, so I gave up on them and just bolded the names of sections/headings.

### CODE REVIEW

**Comment by smile13241324** [comment]
Thanks for the hard work, I will have a look.

As to 1. normally the layers do not tell that you need to have syntax checking layer or auto complete layer installed, there is a general user manual which we expect new comers to read first there the whole system of syntax checking layer and auto complete layer is described in detail.

The individual layers should not repeat information already written down in a general document, especially for maintenance reasons.

**Comment by Martinsos** [comment]
> Thanks for the hard work, I will have a look.
> 
> As to 1. normally the layers do not tell that you need to have syntax checking layer or auto complete layer installed, there is a general user manual which we expect new comers to read first there the whole system of syntax checking layer and auto complete layer is described in detail.
> 
> The individual layers should not repeat information already written down in a general document, especially for maintenance reasons.

Makes sense, the more we duplicate information the harder it is to keep it in sync.

Still, I can say from my experience: I went through most of the spacemacs docs, and still I couldn't even remember there was syntax-checking or auto-complete layer once I got to read Haskell layer README! So it is probably useful to mention at least some things, maybe those that don't change quickly. But I will let you make the final decision on that!

File: layers/+lang/haskell/README.org
Comment (smile13241324) [inline layers/+lang/haskell/README.org:157]: Well the line about LSP not using flycheck is not right, actually there is a specific checker registered within the flycheck system which nowadays is called `lsp` not `lsp-ui` anymore (latest version). However that's still the flycheck system so the same setup instructions apply.

To elaborate a bit more, the traditional emacs system is to register multiple checkers with flycheck which then run sequentially. This is normally slow and requires a lot of setup within emacs. Lsp-mode in contrary passes the job of maintaining linter threads to the external LSP server allowing them to run in parallel. For the emacs user this is visible by that there is only one checker registered called `lsp`. Normally this is ok, however if the server is not well enough designed to be able to run multiple linters some layers have tried to register a second checker after `lsp` this however is very difficult to get right and again is very slow. The right approach in this case is to open a PR for the LSP server to support that specific linter.

File: layers/+lang/haskell/README.org
Comment (smile13241324) [inline layers/+lang/haskell/README.org:321]: I am not sure I like this section. 

Basically it should list all available checkers, which are all for flycheck. This however should not list haskell-mode interactive as this is your repl, which may also have syntax checking in place but is not strictly working on your code files.

So my suggestion would be to remove haskell-mode interactive from this section as well as remove the (via .*) block.

File: layers/+lang/haskell/README.org
Comment (smile13241324) [inline layers/+lang/haskell/README.org:364]: We should remove the section about `flymake` this is horribly outdated.
In addition if we keep this troubleshooting subject, we should move it to the FAQ

File: layers/+lang/haskell/README.org
Comment (smile13241324) [inline layers/+lang/haskell/README.org:352]: Can you double check whether this is still true? Basically standard emacs next error and previous error should work in this buffer too.

File: layers/+lang/haskell/README.org
Comment (smile13241324) [inline layers/+lang/haskell/README.org:159]: Can you verify that links as headlines work in org and markdown? I want to avoid issues with the document conversion beforehand.

**Review by smile13241324 — CHANGES_REQUESTED** [review]
@Martinsos all in all this is a decent rewrite of most of the layer documentation.

I have given it a good read and found one or two points you should check and some smaller ones you should change, before we can merge this into develop.

File: layers/+lang/haskell/README.org
Comment (Martinsos) [reply layers/+lang/haskell/README.org:157]: Thanks for detailed explanation! But from what I read / understood, LSP errors don't appear in flycheck's error list? Or do they then? I would try it myself but I don't have HIE/HLS set up at the moment, might get to it during the weekend but certainly not sooner.

File: layers/+lang/haskell/README.org
Comment (Martinsos) [reply layers/+lang/haskell/README.org:364]: Agreed, I didn't know much about flymake so I didn't want to touch it at the moment. Troubleshooting part -> it does not seem to bring any new information so I will just remove it.

File: layers/+lang/haskell/README.org
Comment (Martinsos) [reply layers/+lang/haskell/README.org:321]: Well I have to admit I don't feel like I have enough knowledge to rewrite this section on my own, and I am not sure I 100% got what you are saying.

Are you proposing that we remove the first part completely, starting with "There are multiple" and ending with "- haskell-mode interactive"? And then that we also remove "** Interactive haskell-mode" section?

My thoughts:
1. I agree this section should be updated :D.
2. If we remove interactive-haskell mode section, are we losing any information? I don't see any other place mentioning error navigation shortcuts that are mentioned here (=haskell-goto-next-error= (~M-n~) and the other one), should we document those somewhere else? Under keybindings probably, but I am not sure if that should go under haskell mode or REPL hm. I could investigate that further, read more about haskell-mode, see if they are valid there also and so on.
3. If we remove that first section and interactive haskell section, not much is left in the Syntax checking section, right? Flycheck section talks about flycheck, which is something that could be looked up in syntax-checking layer docs. It mentions different checkers, but that feels more like implementation detail, why would use need to know explicitly about this? And finally it mentions HLint, but again it does not say anything that can't be looked up in Hlint documentation. So maybe just removing the whole section would make sense? Or making it much much shorter -> but what would you leave in it then?

File: layers/+lang/haskell/README.org
Comment (Martinsos) [reply layers/+lang/haskell/README.org:352]: Hmmm intersting, so I tried it out and what I got is that haskell-goto-prev-error does not work (says No further notes from Haskell compiler) while standard emacs next error and previous error work!
Also, haskell-goto-prev-error does not respond to M-n in my case, it seems like it is not connected to that key binding.

So maybe we should just remove this section completely, the one with haskell interactive mode, as you mentioned?

File: layers/+lang/haskell/README.org
Comment (Martinsos) [reply layers/+lang/haskell/README.org:159]: I can confirm that they work!

File: layers/+lang/haskell/README.org
Comment (smile13241324) [reply layers/+lang/haskell/README.org]: Errors from lsp are registered as flycheck errors, as are all errors you see in any buffer (only if it is using very obscure or veeeerrryyy old code for this).

File: layers/+lang/haskell/README.org
Comment (smile13241324) [reply layers/+lang/haskell/README.org:319]: Good point, no idea yet, lets keep it as it is for now, we can later revise it again.

File: layers/+lang/haskell/README.org
Comment (smile13241324) [reply layers/+lang/haskell/README.org:157]: This is already happening, for the spacemacs websites, especially for develop.spacemacs.org, for this org is converted to markdown. You can check if this still works by running `spacemacs/publish-doc` in `core/core-documentation.el`. This should not fail for the haskell layer.

**Comment by smile13241324** [comment]
Just had another look, I think now only two points are open before we can merge it. Please check the unresolved discussion threads and the we can finally merge it.

File: layers/+lang/haskell/README.org
Comment (Martinsos) [reply layers/+lang/haskell/README.org:157]: Ok makes sense! I actually just moved links to the text, instead of headers, I think it is just eaiser that way. I did run spacemacs/publish-doc from core/core-documentation.el, but after I moved the links from headers, and it generated a lot of stuff and I haven't seen any errors reported. However maybe I missed some error there was a lot of logs, and I wasn't sure where docs got generated so I didn't inspect them. Anyway this should solve it!

File: layers/+lang/haskell/README.org
Comment (Martinsos) [reply layers/+lang/haskell/README.org:319]: Ok, just to confirm -> I left syntax checking section as it was, but removed interactive haskell.

**Comment by Martinsos** [comment]
Ok this should be it! I squashed all commits and took care of all unresolved comments. I also undid resolution of one previously resolved comment, just to confirm that I understood right, which leaves us at 3 unersolved comments right now. If those are all ok, you can merge.

**Review by smile13241324 — APPROVED** [review]

**Comment by smile13241324** [comment]
Thank you for contributing to spacemacs 💜, I have cherry picked your commit into develop.

**Pull request merged**
